packages/dds/merge-tree/docs/Attribution.md

```diff
@@ -0,0 +1,378 @@
+
+### Compaction of similar information
+
+One primary motivation for supporting attribution information natively in the framework is the potential to reduce redundant attribution information in snapshots.
+There are two obvious ways data is redundant: user information gets repeatedly inlined into `JSON.stringify`d content, and various sets of ops all likely have
+virtually the same attribution information (same user, perhaps a slightly different timestamp).
+Ops that have closer together sequence number are more likely to contain such redundant information, as users tend to edit documents in bursts.
+This suggests a few strategies for keeping a compact format (either only on serialization or in-memory as well):
+
+1. Intern user objects
+2. Intern attribution objects
+3. Allow "equivalent timestamp" policy injection: it's unlikely any app needs millisecond or better accuracy on the server ack timestamp for attribution purposes.
+  There should be a configurable policy for how timestamps get binned. Basic implementations could bin on a fixed cadence, but for even more compact files a dynamic bin size policy with larger bins for less recent data could also give a reasonable user experience
+4. If a range of sequence numbers all have the same attribution information, store it as such
+
+#### Interning
+
+Rather than repeatedly serialize the same information in the snapshot format, we can internally add a level of indirection to the `user` field, the entire
+attribution object, or both.
+This optimization would be entirely transparent to the public API: whatever snapshot/in-memory format we use, we'd always convert to `AttributionInfo` before
+returning the information for a given seq to the DDS/application.
+
+Interfaces might look like this, with exported properties being those visible to an application:
+
+```typescript
+export interface AttributionInfo {
+  user: IUser;
+  timestamp: number;
+}
+
+export interface IAttributor {
+  getAttributionInfo(seq: number): AttributionInfo;
+}
+
+type InternedRef = number & { readonly InternedRef: 'e86840d8-8384-450c-b0e3-9a2855ba2d21'};
+
+interface ObjectInterner {
+  getOrCreateRef(obj: Jsonable): InternedRef;
+  getObject(id: InternedRef): Jsonable;
+  getSerializable(): Jsonable;
+}
+
+interface CompactAttributionInfo {
+  userRef: InternedRef;
+  timestamp: number;
+}
+
+// Concrete types for a particular `Attributor` implementation
+interface SerializedAttributor {
+  interner: Jsonable /* result of calling getSerializable() on an ObjectInterner */
+  lookup: {
+    [seq: number]: InternedRef /* to CompactAttributionInfo */ | CompactAttributionInfo
+  }
+}
+```
+
+Another way to accomplish similar gains without the need to compare `IUser`s would be to maintain a historical `clientId -> IUser` lookup and use `clientId` in place of `userRef`.
```

Comment:
I would move away from clientId. it really doesn't tell us much, as it changes all the time on reconnect and such. IUser requires a user id, but even that is a bit large. I think we can actually leverage the sequence number of the first join op for a user with edits, this is a unique and small id

Reply:
we might use client id for a dds to get a reference to the user info for an op in the collab window, so that it can get a reference to the user info

Reply:
we could also consider the unique id stuff @noencke is looking at. particularly at managing the local to acked transition

Reply:
i also know @andre4i is looking at removing join ops in some cases that use custom servers. that will basically make any kind of attribution impossible. maybe that this is fine as it will be pay to play, but something to keep in mind.

Reply:
This whole scheme does have some similarities with the ID compression scheme. We actually built an attribution scheme into the legacy SharedTree's ID compression code since we could re-use the table we already had for IDs. It maps directly from "content ID" -> "user". This doc is proposing splitting that mapping up into two parts: "content ID" -> "sequence number" and then "sequence number" -> "user". The latter is accomplished by the runtime, and the former is accomplished by the DDS. But the former could _also_ be a runtime service for DDS_s; in fact, @justus-camp is investigating this sprint what we could do to lift the ID Compression scheme up to the runtime level as a service. We could at that time also build in an API to the ID compression that gives the sequence number that created an ID. Then both of these services could work together: you use the id generator service to give IDs to your content, you can query it later to get the sequence number, and then you can use the attribution service to get the user for that sequence number. Both services have some similar requirements (for example, you aren't allowed to serialize the objects they give you without first going through some kind of special serializer), and that could help them feel familiar to users and reduce the concept count for picking them (i.e. once you learn the "gotchas" of the attribution API you also know the "gotchas" of the ID compression API). 

Another option would be to compose the ID Compressor API over the attribution API and have the ID compressor also provide an "ID -> user" API (which under the hood uses the attribution service). Then developers using the ID compressor API don't even have to know about the attribution service directly, but if they don't want ID compression we leave the attribution API exposed for direct use and they can maintain their own "content" -> "sequence number" table.

Reply:
> I would move away from clientId. it really doesn't tell us much, as it changes all the time on reconnect and such. IUser requires a user id, but even that is a bit large.

yeah, I think storing the user info directly is probably the better choice here anyway. I mostly wanted to call out here that user objects aren't really comparable/serializable by default right now as they're a point of extensibility. I don't think it will be an issue in practice (most representations people use are probably already JSON.stringify-able without issue and worst-case we could string compare that version).

>  I think we can actually leverage the sequence number of the first join op for a user with edits, this is a unique and small id. it will be possible for this to change, but only in the case where a user info gets gc'd, and the next join + referenced edit will re-add them at their new seq.

It's an interesting idea, but what benefit does this scheme give us? Seems like dictionary compression of the user field should already reduce size enough here.

RE: similarities with a more generalized ID compression:

> This doc is proposing splitting that mapping up into two parts: "content ID" -> "sequence number" and then "sequence number" -> "user".

I'd challenge this a bit. It's not a general feature of DDSes that each piece of content has a unique ID, that's true for SharedTree but not true for SharedString, Matrix, etc. Generally we should be starting with content and not some ID. It is an interesting point though, that rather than make DDSes responsible for extracting a `seq` from the content, they could extract one or more IDs associated with that content that an IdCompressor could attribute to a user.

(I think it is important that the design for attribution allows possibility for multiple-authoring, which is why I mention potentially more than 1 ID)

One relatively clean way to conceptualize this would be for each user to use their id compressor to mint an id for each op they generate, which gives every sequence number an id that could be referenced by DDSes that don't explicitly give ids to their content. This still gives fine-grained attribution detail for DDSes that don't have fine granularity on their content.

Reply:
Sorry I was kind of misleading there; by "content ID" -> "sequence number" I simply mean that a DDS must have some way to get a sequence number from a piece of content, not that that content necessarily is associated with a number/string, etc. that we'd call an "ID". 

I really like your idea of using the compressor to make an ID for each op. The compressor can mint IDs for anything, not just content!

Reply:
having something small, like a join seq integer for the id of a user makes it cheaper for the dds to store so it so the dds can look up a user for an op. this is premised on the op that dds themselves store op data (seq, timestamp, etc) and the higher layers only store use info.

As for user info, the IUser object already must be serializable as it comes over the wire in a join op, and is serialized in the quorum snapshot. we only require and id on that object, but most servers add more properties